src/components/CodeOverview/index.spec.tsx

```diff
@@ -0,0 +1,452 @@
+    const rowHeight = 10;
+    const rowTopPadding = 2;
+
+    const root = render({
+      rowHeight,
+      rowTopPadding,
+      content: generateFileLines({ count: 3 }).join('\n'),
+    });
+    root.setState({ overviewHeight: 400 });
+
+    const innerRoot = renderWithLinterProvider({ root });
+
+    const link1 = innerRoot.find(Link).at(0);
+
+    expect(link1).toHaveStyle('height', `${rowHeight}px`);
+    // The first line should not have top padding.
+    expect(link1).toHaveStyle('paddingTop', undefined);
+
+    const link2 = innerRoot.find(Link).at(1);
+
+    expect(link2).toHaveStyle('height', `${rowHeight}px`);
+    expect(link2).toHaveStyle('paddingTop', `${rowTopPadding}px`);
+
+    const link3 = innerRoot.find(Link).at(2);
+
+    expect(link3).toHaveStyle('height', `${rowHeight}px`);
+    expect(link3).toHaveStyle('paddingTop', `${rowTopPadding}px`);
+  });
+
+  it('sets overview styles', () => {
+    const overviewPadding = 7;
+
+    const root = render({
+      overviewPadding,
+      content: generateFileLines({ count: 3 }).join('\n'),
+    });
+
+    const innerRoot = renderWithLinterProvider({ root });
+
+    expect(innerRoot.find(`.${styles.CodeOverview}`)).toHaveStyle(
+      'padding',
+      `${overviewPadding}px`,
+    );
+  });
+
+  it('renders links for an empty line', () => {
+    const contentLines = ['// Example code comment'];
+
+    const location = createFakeLocation();
+    const root = render({ content: contentLines.join('\n'), location });
+    root.setState({ overviewHeight: 400 });
+
+    const innerRoot = renderWithLinterProvider({ root });
+
+    // Lines 2 and beyond will be empty. Check the first empty line.
+    const link = innerRoot.find(Link).at(1);
+
+    expect(link).toHaveProp('to', {
+      ...location,
+      hash: '#',
+    });
+
+    expect(link).toHaveProp('title', '');
+  });
+
+  it('handles empty content', () => {
+    const content = '';
+    const allLineShapes = generateLineShapes([content]);
+
+    const root = render({ content });
+    root.setState({ overviewHeight: 400 });
+
+    const innerRoot = renderWithLinterProvider({ root });
+
+    const lineShapes = innerRoot.find(CodeLineShapes);
+    expect(lineShapes.at(0)).toHaveProp('lineShapes', allLineShapes[0]);
+    expect(lineShapes).toHaveLength(1);
+  });
+
+  it('renders CodeLineShapes for all lines', () => {
+    const contentLines = generateFileLines({ count: 3 });
+    const allLineShapes = generateLineShapes(contentLines);
+
+    const root = render({ content: contentLines.join('\n') });
+    root.setState({ overviewHeight: 400 });
+
+    const innerRoot = renderWithLinterProvider({ root });
+
+    const lineShapes = innerRoot.find(CodeLineShapes);
+
+    expect(lineShapes.at(0)).toHaveProp('lineShapes', allLineShapes[0]);
+    expect(lineShapes.at(1)).toHaveProp('lineShapes', allLineShapes[1]);
+    expect(lineShapes.at(2)).toHaveProp('lineShapes', allLineShapes[2]);
+
+    expect(lineShapes).toHaveLength(3);
```

Comment:
Nit: I would expect this assertion to appear above the ones for the individual `lineShapes`, because if there are less than 3 `lineShapes` then the test should fail and give a more descriptive message than it would if it encountered a `toHaveProp` for a nonexistent line.

Reply:
The failure message for `toHaveLength()` is not a good one so I try to put it at the end when possible. Additionally, in this case, it's only there to catch an unexpected amount of rows so it's secondary to the main assertions.